### ISSUE TITLE
validate localhost

### ISSUE THREAD

**Issue by jfoclpf** [zrrrzzt/html-validator#105]
Does it validate html served by `localhost`?

```
const validator = require('html-validator')
const options = {
 url: 'http://localhost',
 format: 'text'
}
```

Most likely no, since it uses W3 online tool, but how do I do that?
I want to test my html before sending it to the www!

Thanks

**Comment by jfoclpf** [comment]
Oh, I see that we can read files, so it does work offline :)

```js
const validator = require('html-validator')
const fs = require('fs')
var options = {
  format: 'text'
}
 
fs.readFile( 'file-to-validate.html', 'utf8', (err, html) => {
  if (err) {
    throw err;
  }
  
  options.data = html
 
  validator(options)
    .then((data) => {
      console.log(data)
    })
    .catch((error) => {
      console.error(error)
    })
})
```

Great! Just a tip, in your readme file maybe put this example first, because it is not clear at first glance what's the difference between this package and the w3 online tool. I mean, it's not clear at first glance that we can use this tool offline.

Thanks

**Comment by zrrrzzt** [comment]
You cannot use this offline. It's just a wrapper for the online validator. By reading the file you post the content to w3c instead of the url.

Regarding localhost you are right. You could validate the content of a file from you local environment.
Another approach for local development could be [ngrok](https://ngrok.com/) or [localtunnel](https://github.com/localtunnel/localtunnel#readme)

**Comment by jfoclpf** [comment]
Thanks, therefore, if I understood correctly

1. the server must be online
2. but the html file/string may be stored and/or generated on my local machine

thank you

**Comment by jfoclpf** [comment]
still regarding this issue, it seems I cannot access directly `localhost` even if I'm online

```js
      var url = 'http://localhost:' + settings.HTTPport + '/PT'
      console.log('Validating html on:', url)

      const options = {
        url: url,
        format: 'text'
      }

      validator(options, (error, data) => {
        if (error) {
          callback(Error(error))
        } else {
          console.log(data)
          callback()
        }
      })
```

returns

```
Validating html on: http://localhost:3026/PT
IO Error: Connect to localhost:3026 [localhost/127.0.0.1, localhost/0:0:0:0:0:0:0:1] failed: Connection refused (Connection refused)
The result cannot be determined due to a non-document-error.
```

and though I can access `http://localhost:3026/PT` on my browser

**Comment by zrrrzzt** [comment]
You can access localhost on your browser but the validator service can't :-)
That's why you'll have to use a service like [ngrok](https://ngrok.com/) or [localtunnel](https://github.com/localtunnel/localtunnel#readme)

Here's an example using a ngrok wrapper for Node https://www.npmjs.com/package/ngrok

```JavaScript
(async () => {
  const ngrok = require('ngrok')
  const validator = require('html-validator')
  const ngrokUrl = await ngrok.connect(settings.HTTPport)
  const url = `${ngrokUrl}/PT`
  console.log('Validating html on:', url)
  const options = {
     url: url,
     format: 'text'
   }
  
   try {
    const data = await validator(options)
    console.log(data)
  } catch (error) {
    console.error(error)
  }
})()
```

Disclaimer: very untested code

**Comment by jfoclpf** [comment]
Thank you very much. In any case, just FYI I used another method without the need to use external services. I used the module `request` and pass the result to `option.data`

I tested now and it works :)

```js
var url = 'http://localhost:' + settings.HTTPport + '/PT'
request({ uri: url }, function (err, response, body) {
  if (err) {
    callback(Error(err))
  } else {
    // console.log(body)
    let options = {
      format: 'text',
      data: body
    }
    validator(options)
      .then((result) => {
        if (result.includes('Error')) {
          console.log('Error found on ' + url , result)
        } else {
          console.log(result)
        }
      })
      .catch((err) => {
        console.log(err)
        callback(Error(err))
      })
  }
})
```

**Comment by zrrrzzt** [comment]
Good to see you solved it. And yes, if you post the data yourself you have no need for external services to be able to reach your localhost.

Regarding error parsing: I think checking for Error or There were errors is what you do for formats other than json. If you use json you'll need to check type for each message. Take a look at https://github.com/zrrrzzt/html-validator-cli/blob/master/index.js to see how I solved it for the cli of this module

**Comment by zrrrzzt** [comment]
just released v4.0.0 and implemented isLocal as a flag for validating from localhost.

```JavaScript
(async () => {
  const validator = require('html-validator')
  const options = {
    url: 'http://localhost:' + settings.HTTPport + '/PT',
    format: 'text',
    isLocal: true
  }
  
  try {
    const result = await validator(options)
    console.log(result)
  } catch (error) {
    console.error(error)
  }
})()
```

**Comment by jfoclpf** [comment]
Thanks a lot, I'll update accordingly
